### DISCUSSION TITLE
Libraries for Fabric control plane

### DISCUSSION THREAD

**Discussion by rachit1arora** [fybrik/fybrik#786] [Ideas]
The purpose of discussion is to discuss the feasibility of deployment models for fybrik. 

Current model expect to create end user to submit fybrikApplication a custom resource to start the enforcement flow. Once a fybrik application is created it creates various other custom resources to achieve the goal. 
It is expected that a user has access to kube API server to submit these applications. 

Many organisations are very particular to open up kube API server  or even many users do not know Kubernetes. Thus it will be beneficial if we can provide a deployment config which allow to submit request for fybrik application and control plane orchestration without using Kubernetes Custom Resources. 

One option is to have our fybrik application, plotter components par of a go library and the fybrik application can be invoked through API to start the orchestration flow and control plane generate out as blueprint which again can be a json or another Kubernetes Custom resource based on deployment config.

**Comment by simanadler** [comment]
The sample GUI implementation provided has a REST interface that enables the creation of the Fybrik application, so the user doesn't need to know kubernetes.  The GUI server calls kubernetes.  As is currently configured though, the kubernetes api is called from within the user namespace.

**Comment by roee88** [comment]
There are two items here if I understand correctly:
1. Manage FybrikApplications without access to Kubernetes 
2. Support a model without using CRDs for fybrik application and plotter

This discussion mixes the two. Can you confirm that you meant both @rachit1arora? Just to focus the discussion

**Comment by rachit1arora** [comment]
@roee88 This Discussions is primarily to for second point `Support a model without using CRDs for fybrik application and plotter` and first point is primarily a motivator for the requirement as many organisations want to avoid Kube api server access for end users as well as internal processes

**Reply by simanadler** [reply]
@rachit1arora Are there other motivators for not using CRDs for fybrik application and plotter?  If avoiding the direct use of the kube api server, then that is addressed by REST APIs which hide the kubernetes access from the user.

**Reply by roee88** [reply]
@simanadler it's not strictly about direct use. My understanding is that for use cases like public cloud services avoiding the shared etcd is desired.
